src/classes/Shapes/CapsuleShape.js

```diff
@@ -1,425 +1,188 @@
-
-						if ( t4 < 0.0 ) { // segment is pointing away from the top sphere
-							return null;
-						}
-
-						if ( t4 <= length ) { // intersection is inside of the segment
-							intersection = getIntersectionFromDirection( start, t4 );
-						} else { // intersection is after segment ends
-							return null;
-						}
-					} else if ( y2 < -this.half_height ) { // line intersects cylinder below capsule bottom
-
-						a += direction.y * direction.y;
-						c += ( start.y + this.half_height ) * ( start.y + this.half_height );
-						k += direction.y * ( start.y + this.half_height );
-						discr = k * k - a * c;
-
-						if ( discr <= 0.0 ) { // line doesn't intersect bottom sphere or has single intersection point, that should never happen
-							return null;
-						}
-
-						discr_sqrt = Math.sqrt( discr );
-						t4 = ( -k + discr_sqrt ) / a; // line and bottom intersection point, the most distant from the start point
-
-						if ( t4 < 0.0 ) { // segment is pointing away from the bottom sphere
-							return null;
-						}
-
-						if ( t4 <= length ) { // intersection is inside of the segment
-							intersection = getIntersectionFromDirection( start, t4 );
-						} else { // intersection is after segment ends
-							return null;
-						}
-					} else { // line intersects capsule side surface
-						if ( t2 <= length ) { // intersection is inside of the segment
-							intersection = getIntersectionFromDirection( start, t2 );
-						} else { // intersection is after segment ends
-							return null;
-						}
-					}
-				}
-			}
-		}
-
-		intersection.normal.x = intersection.point.x;
-		intersection.normal.z = intersection.point.z;
-		if ( intersection.point.y < -this.half_height ) {
-			intersection.normal.y = intersection.point.y + this.half_height;
-		} else if ( intersection.point.y > this.half_height ) {
-			intersection.normal.y = intersection.point.y - this.half_height;
-		} else {
-			intersection.normal.y = 0.0;
-		}
-		intersection.normal.scale( 1.0 / this.radius );
-		intersection.object = this;
-
-		return intersection;
-	};
-} )();
+Goblin.CapsuleShape.prototype.rayIntersect = ( function() {
+    var pa = new Goblin.Vector3();
+    var pb = new Goblin.Vector3();
+    var rd = new Goblin.Vector3();
+
+    var ba = new Goblin.Vector3();
+    var oa = new Goblin.Vector3();
+    var oc = new Goblin.Vector3();
+
+    /**
+     * Copied from https://www.iquilezles.org/www/articles/intersectors/intersectors.htm, all variable names are preserved.
+     * @param {Goblin.Vector3} ro - ray origin
+     * @param {Goblin.Vector3} rd - ray direction
+     * @param {Goblin.Vector3} pa - center point of the first sphere
+     * @param {Goblin.Vector3} pb - center point of the second sphere
+     * @param {Goblin.Vector3} ra - radius of spheres
+     * @returns {number} t - such as ro + rd * t === intersection point
+     */
+    var rayIntersectInternal = function( ro, rd, pa, pb, ra ) {
```

Comment:
Thank you, sir! Finally, a proper implementation :)

Reply:
my pleasure